File: windows/PabloCompanion.Tests/ViewModels/SubscriptionViewModelTests.cs
using PabloCompanion.Models;
using PabloCompanion.ViewModels;
using PabloCompanion.Services;

namespace PabloCompanion.Tests.ViewModels;

/// <summary>
/// Verifies the 0=unlimited convention for trial_sessions_limit and trial_days_limit.
/// Backend contract: a limit of 0 means no cap — the UI must not show "0 remaining".
/// </summary>
public sealed class SubscriptionViewModelTests
{
    // --- TrialSessionsRemaining ---

    [Fact]
    public void TrialSessionsRemaining_NormalLimit_ReturnsRemainder()
    {
        var vm = MakeVm(sessionsUsed: 8, sessionsLimit: 20);
        Assert.Equal(12, vm.TrialSessionsRemaining);
    }

    [Fact]
    public void TrialSessionsRemaining_ZeroLimit_ReturnsNull()
    {
        // trial_sessions_limit=0 means unlimited — must not produce "0 sessions remaining".
        var vm = MakeVm(sessionsUsed: 5, sessionsLimit: 0);
        Assert.Null(vm.TrialSessionsRemaining);
    }

    [Fact]
    public void TrialSessionsRemaining_NullLimit_ReturnsNull()
    {
        var vm = MakeVm(sessionsUsed: 3, sessionsLimit: null);
        Assert.Null(vm.TrialSessionsRemaining);
    }

    [Fact]
    public void TrialSessionsRemaining_UsedExceedsLimit_ClampsToZero()
    {
        var vm = MakeVm(sessionsUsed: 25, sessionsLimit: 20);
        Assert.Equal(0, vm.TrialSessionsRemaining);
    }

    // --- TrialDaysRemaining ---

    [Fact]
    public void TrialDaysRemaining_NormalLimit_ReturnsApproxRemainder()
    {
        var start = DateTimeOffset.UtcNow.AddDays(-10).ToString("O");
        var vm = MakeVm(daysLimit: 30, trialStart: start);
        var remaining = vm.TrialDaysRemaining;
        Assert.NotNull(remaining);
        Assert.InRange(remaining!.Value, 19, 20);
    }

    [Fact]
    public void TrialDaysRemaining_ZeroLimit_ReturnsNull()
    {
        // trial_days_limit=0 means unlimited — must not produce "0 days left".
        var start = DateTimeOffset.UtcNow.AddDays(-10).ToString("O");
        var vm = MakeVm(daysLimit: 0, trialStart: start);
        Assert.Null(vm.TrialDaysRemaining);
    }

    [Fact]
    public void TrialDaysRemaining_NullLimit_ReturnsNull()
    {
        var start = DateTimeOffset.UtcNow.AddDays(-5).ToString("O");
        var vm = MakeVm(daysLimit: null, trialStart: start);
        Assert.Null(vm.TrialDaysRemaining);
    }

    // --- Helpers ---

    private static SubscriptionViewModel MakeVm(
        int? sessionsUsed = null,
        int? sessionsLimit = null,
        int? daysLimit = null,
        string? trialStart = null)
    {
        var vm = new SubscriptionViewModel(null!);
        vm.Info = new SubscriptionInfo(
            Status: "trial",
            Plan: "solo",
            TrialSessionsUsed: sessionsUsed,
            TrialSessionsLimit: sessionsLimit,
            TrialDaysLimit: daysLimit,
            TrialStart: trialStart,
            GraceExtensionAvailable: false,
            GraceExtensionExpiresAt: null);
        return vm;
    }
}

File: mac/PabloCompanionTests/SubscriptionViewModelTests.swift
import Foundation
@testable import Pablo
import Testing

@Suite("SubscriptionViewModel trial limits")
@MainActor
struct SubscriptionViewModelTrialTests {

    // MARK: - trialSessionsRemaining

    @Test func sessionsRemaining_normalLimit() {
        let vm = SubscriptionViewModel()
        vm.subscriptionInfo = makeInfo(sessionsUsed: 8, sessionsLimit: 20)
        #expect(vm.trialSessionsRemaining == 12)
    }

    @Test func sessionsRemaining_zeroLimitIsUnlimited() {
        // Backend contract: trial_sessions_limit=0 means no session cap.
        let vm = SubscriptionViewModel()
        vm.subscriptionInfo = makeInfo(sessionsUsed: 5, sessionsLimit: 0)
        #expect(vm.trialSessionsRemaining == nil, "limit=0 must return nil (unlimited), not 0")
    }

    @Test func sessionsRemaining_nilLimitReturnsNil() {
        let vm = SubscriptionViewModel()
        vm.subscriptionInfo = makeInfo(sessionsUsed: 3, sessionsLimit: nil)
        #expect(vm.trialSessionsRemaining == nil)
    }

    @Test func sessionsRemaining_clampedToZero() {
        let vm = SubscriptionViewModel()
        vm.subscriptionInfo = makeInfo(sessionsUsed: 25, sessionsLimit: 20)
        #expect(vm.trialSessionsRemaining == 0)
    }

    // MARK: - trialDaysRemaining

    @Test func daysRemaining_normalLimit() throws {
        let vm = SubscriptionViewModel()
        vm.subscriptionInfo = makeInfo(
            daysLimit: 30,
            trialStart: iso8601(daysAgo: 10)
        )
        let remaining = vm.trialDaysRemaining
        #expect(remaining != nil)
        #expect(try #require(remaining) >= 19 && remaining! <= 20, "Expected ~20 days remaining")
    }

    @Test func daysRemaining_zeroLimitIsUnlimited() {
        // Backend contract: trial_days_limit=0 means no time cap.
        let vm = SubscriptionViewModel()
        vm.subscriptionInfo = makeInfo(
            daysLimit: 0,
            trialStart: iso8601(daysAgo: 10)
        )
        #expect(vm.trialDaysRemaining == nil, "daysLimit=0 must return nil (unlimited), not 0")
    }

    @Test func daysRemaining_nilLimitReturnsNil() {
        let vm = SubscriptionViewModel()
        vm.subscriptionInfo = makeInfo(daysLimit: nil, trialStart: iso8601(daysAgo: 5))
        #expect(vm.trialDaysRemaining == nil)
    }

    // MARK: - bannerState with limit=0

    @Test func bannerState_unlimitedSessionsAndDays_showsGenericTrialText() {
        let vm = SubscriptionViewModel()
        vm.subscriptionInfo = makeInfo(
            sessionsUsed: 5, sessionsLimit: 0,
            daysLimit: 0, trialStart: iso8601(daysAgo: 10)
        )
        if case let .trial(sessions, days) = vm.bannerState {
            #expect(sessions == nil, "Unlimited sessions must be nil in banner state")
            #expect(days == nil, "Unlimited days must be nil in banner state")
        } else {
            Issue.record("Expected .trial banner state")
        }
    }

    // MARK: - Helpers

    private func makeInfo(
        sessionsUsed: Int? = nil,
        sessionsLimit: Int? = nil,
        daysLimit: Int? = nil,
        trialStart: String? = nil
    ) -> SubscriptionInfo {
        SubscriptionInfo(
            status: .trial,
            plan: "solo",
            trialSessionsUsed: sessionsUsed,
            trialSessionsLimit: sessionsLimit,
            trialDaysLimit: daysLimit,
            trialStart: trialStart,
            graceExtensionAvailable: false,
            graceExtensionExpiresAt: nil
        )
    }

    private func iso8601(daysAgo: Double) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.string(from: Date().addingTimeInterval(-86400 * daysAgo))
    }
}
